server/webex_store.go

```diff
@@ -0,0 +1,267 @@
+
+}
+
+// loadWebexSession retrieves the webex session if present from the KV store
+func (p *Plugin) loadWebexSession(userID string) (*WebexOAuthSession, error) {
+
+	var session *WebexOAuthSession
+
+	key := fmt.Sprintf("%s%s", WebexOAuthSessionKey, userID)
+
+	data, appErr := p.API.KVGet(key)
+	if appErr != nil {
+		p.API.LogError(
+			"Error retrieving webex session",
+			"user_id", userID,
+			"key", key,
+			"error", appErr.Error(),
+		)
+		return session, appErr
+	}
+
+	if len(data) == 0 {
+		return session, ErrWebexSessionNotFound
+	}
+
+	err := json.Unmarshal(data, &session)
+	if err != nil {
+		return session, err
+	}
+
+	config := p.getConfiguration()
+
+	if time.Now().UTC().After(session.Token.Expiry) {
+		kvErr := p.API.KVDelete(key)
+		if kvErr != nil {
+			return nil, kvErr
+		}
+		return nil, ErrWebexSessionExpired
+	}
+
+	accessToken, err := decrypt([]byte(config.EncryptionKey), session.Token.AccessToken)
+	if err != nil {
+		return session, err
+	}
+
+	session.Token.AccessToken = accessToken
+
+	// Decrypt refresh token if it's present
+	if session.Token.RefreshToken != "" {
+
+		refreshToken, err := decrypt([]byte(config.EncryptionKey), session.Token.RefreshToken)
+		if err != nil {
+			return session, err
+		}
+
+		session.Token.RefreshToken = refreshToken
+	}
+
+	return session, nil
+}
+
+// getWebexUserInfo retrieves the webex user info from the KV store and
+// if not present will request from the webex API and subsequently save
+// it to the KV store if successful
+func (p *Plugin) getWebexUserInfo(userID string) (*WebexUserInfo, error) {
+
+	webexUser, loadErr := p.loadWebexUser(userID)
+
+	switch {
+	case loadErr == ErrWebexUserNotFound:
+
+		session, err := p.loadWebexSession(userID)
+		if err != nil {
+			p.API.LogError("error looking up session", "error", err.Error())
+			return webexUser, err
+		}
+
+		webex, err := NewWebexClient(session.Token.AccessToken)
+		if err != nil {
+			p.API.LogError("Error creating new webex client", "error", err.Error())
+			return webexUser, err
+		}
+
+		person, _, err := webex.People.GetMe() // don't need resp so supressing it
+		if err != nil {
+			p.API.LogError("Error calling people.GetMe()", "error", err.Error())
+			return webexUser, err
+		}
+
+		webexUser.FromWebexPerson(person)
+
+		err = p.storeWebexUser(userID, webexUser)
+		if err != nil {
+			p.API.LogError("Error saving webex user", "error", err.Error())
+			return webexUser, err
+		}
+
+	case (loadErr != nil):
+		return webexUser, loadErr
+	}
+
+	return webexUser, nil
+
+}
+
+// loadWebexUser loads the webex user info from the KV store if present
+func (p *Plugin) loadWebexUser(userID string) (*WebexUserInfo, error) {
+
+	var webexUser *WebexUserInfo
+
+	key := fmt.Sprintf("%s%s", WebexUserKey, userID)
+
+	data, appErr := p.API.KVGet(key)
+	if appErr != nil {
+		p.API.LogError(
+			"Error retrieving webex user info",
+			"user_id", userID,
+			"key", key,
+			"error", appErr,
+		)
+		return webexUser, appErr
```

Comment:
As soon we have the `*WebexUserInfo` as response, normally you put in this cases `return nil, appErr` which is super clear in the intention, but `return webexUser, appErr` still makes me think you still want to return some kind of "partial result" and an error.

Reply:
Agreed, updated.